File: app/src/main/java/app/we/go/emojidraw/arch/mvp/BasePresenter.java
package app.we.go.emojidraw.arch.mvp;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import io.reactivex.disposables.CompositeDisposable;


public class BasePresenter<V extends MVP.View> implements MVP.Presenter<V> {

    @Nullable
    protected V view;

    final protected CompositeDisposable disposables = new CompositeDisposable();

    @Override
    public final void bindView(@NonNull final V view) {
        this.view = view;
        init();
    }

    protected void init() {

    }

    @Override
    public void unbindView() {
        view = null;
        disposables.clear();
    }

    @Override
    public boolean isViewBound() {
        return (view != null);
    }
}

File: app/src/main/java/app/we/go/emojidraw/arch/mvp/BasePresenter.kt
package app.we.go.emojidraw.arch.mvp

import io.reactivex.disposables.CompositeDisposable


open abstract class BasePresenter<V : MVP.View> : MVP.Presenter<V> {

    protected var view: V? = null

    protected val disposables = CompositeDisposable()

    override fun bindView(view: V) {
        this.view = view
        initialize()
    }

    protected open fun initialize() {
        // no-op
    }

    override fun unbindView() {
        view = null
        disposables.clear()
    }

    override fun isViewBound(): Boolean {
        return view != null
    }
}
